### PR TITLE
fix: JS_NewFloat64 -O3 not support Infinity NaN

### LINKED ISSUE #399: osx clang use -O3 parseFloat('Infinity') == 3
## reproduce
```sh
> git clone --recursive https://github.com/saghul/txiki.js --shallow-submodules && cd txiki.js
> cmake -S . -B build -GNinja
> cmake --build build
> echo "console.log(parseFloat('Infinity'), parseFloat("-Infinity"))" > 1.js
> ./build/tjs run 1.js
3 3
```
## minimal c reproduction

**1.c**
``` c
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include <math.h>

int main() {
    double d = INFINITY;
    int32_t val;
    union {
        double d;
        uint64_t u;
    } u, t;
    u.d = d;
    val = (int32_t)d;
    t.d = val;
    bool ok = (u.u == t.u);
    printf("%d\n", ok);
}
```
**use -O3 reproduction**
```sh
> cc 1.c
> ./a.out
0
> cc -O3 1.c
> ./a.out
-1273920960
```

[u.u == t.u](https://github.com/saghul/txiki.js/blob/master/deps/quickjs/include/quickjs.h#L558)

## solution
``` patch
From 7c3523cb349ca49572b6baeba061cff76fb634cb Mon Sep 17 00:00:00 2001
From: zeromake <a390720046@gmail.com>
Date: Wed, 27 Sep 2023 18:53:17 +0800
Subject: [PATCH] fix: -O3 nan inf float

---
 deps/quickjs/include/quickjs.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/deps/quickjs/include/quickjs.h b/deps/quickjs/include/quickjs.h
index 0908520..c2af56e 100644
--- a/deps/quickjs/include/quickjs.h
+++ b/deps/quickjs/include/quickjs.h
@@ -27,6 +27,7 @@
 
 #include <stdio.h>
 #include <stdint.h>
+#include <math.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -544,6 +545,9 @@ JSValue JS_NewBigUint64(JSContext *ctx, uint64_t v);
 
 static js_force_inline JSValue JS_NewFloat64(JSContext *ctx, double d)
 {
+    if (isinf(d) || isnan(d)) {
+        return __JS_NewFloat64(ctx, d);
+    }
     JSValue v;
     int32_t val;
     union {
-- 
2.42.0
```

### PR DESCRIPTION
**Pull request by zeromake**
close #399

### CODE REVIEW

**Comment by saghul** [comment]
Please don't change the CMake version, it's not related to your change AFAICT.

**Comment by zeromake** [comment]
@saghul CMake version change is reset

**Comment by zeromake** [comment]
@saghul Incidentally cmake latest version is `3.27`, in deps/quickjs/CMakeLists.txt `cmake_minimum_required(VERSION 3.4)` what's going on

**Comment by zeromake** [comment]
@saghul 
https://github.com/saghul/txiki.js/actions/runs/6333568907/job/17204665047
```js
// (25).toExponential(0) on mingw is 2e+1
assert.eq((25).toExponential(0), "3e+1");
```
I'll find a Windows system debugging later

File: deps/quickjs/include/quickjs.h [hunk 1]
```diff
@@ -544,6 +545,9 @@ JSValue JS_NewBigUint64(JSContext *ctx, uint64_t v);
 
 static js_force_inline JSValue JS_NewFloat64(JSContext *ctx, double d)
 {
+    if (isinf(d) || isnan(d)) {
```
Comment (saghul) [inline deps/quickjs/include/quickjs.h:548]: Can you add an unlikely here?

Comment (zeromake) [reply deps/quickjs/include/quickjs.h]: @saghul Added

File: deps/quickjs/include/quickjs.h [hunk 2]
```diff
@@ -544,6 +545,9 @@ JSValue JS_NewBigUint64(JSContext *ctx, uint64_t v);
 
 static js_force_inline JSValue JS_NewFloat64(JSContext *ctx, double d)
 {
+    if (unlikely(isinf(d) || isnan(d))) {
```
Comment (saghul) [inline deps/quickjs/include/quickjs.h:548]:
```suggestion
    if (js_unlikely(isinf(d) || isnan(d))) {
```

File: deps/quickjs/CMakeLists.txt [hunk 3]
```diff
@@ -52,3 +52,9 @@ target_include_directories(qjs PRIVATE src)
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
     target_link_libraries(qjs atomic)
 endif()
+
+if(MINGW)
+    target_compile_definitions(qjs PRIVATE
+        CONFIG_PRINTF_RNDN
```
Comment (zeromake) [inline deps/quickjs/CMakeLists.txt:58]: @saghul 
I find quickjs has definde `CONFIG_PRINTF_RNDN`, but my windows not has curl lib, so this code is not testing

**Comment by zeromake** [comment]
rounding snprintf on windows llvm-mingw unlike Linux
``` c
/* fesetround example */
#include <stdio.h>    /* printf */
#include <fenv.h>    /* fesetround, FE_* */
#include <math.h>    /* rint */

int main()
{
    float arr[6] = {28, -28, 25, -25, 24, -24};
    char buf[128] = {0};
    for (int i = 0; i < 6; i++) {
        float v = arr[i];
        printf("--------------------rounding %f:\n", v);
        fesetround(FE_DOWNWARD);
        
        snprintf(buf, 128, "%+.*e", 0, v);
        printf("FE_DOWNWARD: %s\n", buf);
        fesetround(FE_TONEAREST);
        snprintf(buf, 128, "%+.*e", 0, v);
        printf("FE_TONEAREST: %s\n", buf);
        fesetround(FE_TOWARDZERO);
        snprintf(buf, 128, "%+.*e", 0, v);
        printf("FE_TOWARDZERO: %s\n", buf);
        fesetround(FE_UPWARD);
        snprintf(buf, 128, "%+.*e", 0, v);
        printf("FE_UPWARD: %s\n\n", buf);
    }
    return 0;
}

```

**on llvm-mingw**
``` sh
# -O0
--------------------rounding 28.000000:
FE_DOWNWARD: +3e+01
FE_TONEAREST: +3e+01
FE_TOWARDZERO: +3e+01
FE_UPWARD: +3e+01

--------------------rounding -28.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -3e+01
FE_TOWARDZERO: -3e+01
FE_UPWARD: -3e+01

--------------------rounding 25.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +2e+01

--------------------rounding -25.000000:
FE_DOWNWARD: -2e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 24.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +2e+01

--------------------rounding -24.000000:
FE_DOWNWARD: -2e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

# -O3
--------------------rounding 28.000000:
FE_DOWNWARD: +3e+01
FE_TONEAREST: +3e+01
FE_TOWARDZERO: +3e+01
FE_UPWARD: +3e+01

--------------------rounding -28.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -3e+01
FE_TOWARDZERO: -3e+01
FE_UPWARD: -3e+01

--------------------rounding 25.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +2e+01

--------------------rounding -25.000000:
FE_DOWNWARD: -2e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 24.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +2e+01

--------------------rounding -24.000000:
FE_DOWNWARD: -2e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01
```

**on linux gcc**
```sh
# -O0
--------------------rounding 28.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +3e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -28.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -3e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 25.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -25.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 24.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -24.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

# -O3
--------------------rounding 28.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +3e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -28.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -3e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 25.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -25.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01

--------------------rounding 24.000000:
FE_DOWNWARD: +2e+01
FE_TONEAREST: +2e+01
FE_TOWARDZERO: +2e+01
FE_UPWARD: +3e+01

--------------------rounding -24.000000:
FE_DOWNWARD: -3e+01
FE_TONEAREST: -2e+01
FE_TOWARDZERO: -2e+01
FE_UPWARD: -2e+01
```

**Comment by saghul** [comment]
Wow we are going deep here :-) Do you reckon this is a bug in the compiler? Making that particular test dependent on the platform, with a comment and a link to this issue sounds like an acceptable compromise IMHO.

**Comment by zeromake** [comment]
> Wow we are going deep here :-) Do you reckon this is a bug in the compiler? Making that particular test dependent on the platform, with a comment and a link to this issue sounds like an acceptable compromise IMHO.

It doesn't feel too good to handle this pr only for Infinity NaN

**Pull request merged**